llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -719,6 +724,26 @@ void updatePaddingInLastMDNode(LLVMContext &Ctx,
   SCMetadata[Last.first] = MDNode::get(Ctx, MDOps);
 }
 
+/// Function creates 'store' instruction from the given Value @V into
+/// the given Value @Dst.
+/// Note: Types of values Dst and V might differ because of padding bytes
+/// inserted by Clang FE.
+/// For example:
+/// Type of specialization constant might be <{ i32, i8, [ 3 x i8 ] }>, where
+/// the last component are padding bytes.
+/// specialization id in this case could be { i32, i8 } { i32 1, i8 1 }.
+/// As you can see, padding bytes are absent. In order to mitigate this we
+/// perform bitcast from specialization id type to specialization constant
+/// type.
```

Comment:
I wonder how will it work if a padding is inserted in the middle of a struct. **Not** asking to address in this PR, because it is known that we can't handle such structs anyway and therefore it should be addressed as a whole in a separate PR. I have initial set of changes for that [in my fork](https://github.com/AlexeySachkov/llvm/commits/private/asachkov/spec-constants-with-padding)